### PR TITLE
[dpdk] Eliminate header copy

### PR DESCRIPTION
**Pull request by kamleshbhalui**
Dpdk does not support direct header copy.
So, Eliminate temporary copies of header, which generally populated after inlining, if it;s not temporary copy then transform direct header copy to element wise copy.

### CODE REVIEW

File: backends/dpdk/dpdkArch.cpp [hunk 1]
```diff
@@ -2253,4 +2253,63 @@ void CollectAddOnMissTable::postorder(const IR::MethodCallStatement *mcs) {
...
+            replacementMap.insert(std::make_pair(path->path->name.name,
+                                   as->right->to<IR::Member>()));
+            return new IR::EmptyStatement();
+        } else if (auto path = as->right->to<IR::PathExpression>()) {
+            replacementMap.insert(std::make_pair(path->path->name.name,
+                                  as->left->to<IR::Member>()));
+            return  new IR::EmptyStatement();
+        } else {
+            // both are non temporary header instance do element wise copy
+            IR::IndexedVector<IR::StatOrDecl> components;
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkArch.cpp:2275]: This looks wrong, since it won't copy the valid flag.

File: backends/dpdk/dpdkArch.cpp [hunk 2]
```diff
@@ -2253,4 +2253,63 @@ void CollectAddOnMissTable::postorder(const IR::MethodCallStatement *mcs) {
...
+    auto type = typeMap->getType(e);
+    if (type)
+        return type->is<IR::Type_Header>() && !e->is<IR::MethodCallExpression>();
+    return false;
+}
+
+const IR::Node* ElimHeaderCopy::preorder(IR::AssignmentStatement* as) {
+    if (isHeader(as->left) && isHeader(as->right)) {
+        if (auto path = as->left->to<IR::PathExpression>()) {
+            replacementMap.insert(std::make_pair(path->path->name.name,
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkArch.cpp:2266]: what is happening here?

File: backends/dpdk/dpdkArch.h [hunk 3]
```diff
@@ -926,6 +926,20 @@ class CollectErrors : public Inspector {
     }
 };
 
+// Eliminate temporary copies of header, which generally populated after inlining,
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkArch.h:929]: We have a similar pass for copying structures, but it won't really apply here. It handles cases of nested structures.
Does this pass handle headers with structs inside?
This comment should use ///.
Also, please add a P4 code example showing the before and after.

File: backends/dpdk/dpdkArch.cpp [hunk 4]
```diff
@@ -2253,4 +2253,63 @@ void CollectAddOnMissTable::postorder(const IR::MethodCallStatement *mcs) {
...
+}
+
+const IR::Node* ElimHeaderCopy::preorder(IR::MethodCallExpression *me) {
+    auto m = me->method->to<IR::Member>();
+    if (!m)
+        return me;
+    if (!isHeader(m->expr))
+        return me;
+    auto expr = m->expr->to<IR::PathExpression>();
+    if (expr && replacementMap.count(expr->path->name.name)) {
```
Comment (mbudiu-vmw) [inline backends/dpdk/dpdkArch.cpp:2294]: I cannot guess what is happening here, can you please add comments explaining?

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]

File: backends/dpdk/dpdkArch.h (hunk 3 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.h:929]: >Does this pass handle headers with structs inside?
Actually structs inside gets flatten by midend 
i.e.
```
struct foo {
    EthernetAddress dstAddr;
    EthernetAddress srcAddr;
    bit<16>         etherType;
}

header ethernet_t {
    foo f;
    EthernetAddress dstAddr;
    EthernetAddress srcAddr;
    bit<16>         etherType;
}
```

becomes
```
header ethernet_t {
    bit<48> _f_dstAddr0;
    bit<48> _f_srcAddr1;
    bit<16> _f_etherType2;
    bit<48> _dstAddr3;
    bit<48> _srcAddr4;
    bit<16> _etherType5;
}
```

File: backends/dpdk/dpdkArch.cpp (hunk 4 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp:2294]: it's for 
replacing eth_0.setInvalid()
with hdr.ethernet.setInvalid()

File: backends/dpdk/dpdkArch.cpp (hunk 2 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp:2266]: It's for populating replacement map by keeping temporary header name as key and source of assignment statement as value.
i.e.
for below assignment statement
 eth_0 = hdr.ethernet
replacement map will be like
replacementMap["eth_0"] = hdr.ethernet;

later on all uses of eth_0 will be replace with there value in replacementMap.

File: backends/dpdk/dpdkArch.cpp (hunk 1 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp:2275]: done

File: backends/dpdk/dpdkArch.cpp (hunk 4 above)
Comment (kamleshbhalui) [reply backends/dpdk/dpdkArch.cpp:2330]: added

File: backends/dpdk/dpdkArch.cpp (hunk 2 above)
Comment (mbudiu-vmw) [reply backends/dpdk/dpdkArch.cpp:2266]: I would add this information as a comment in the header file where the replacementMap is defined.

**Review by mbudiu-vmw — APPROVED** [review]

**Review by mbudiu-vmw — APPROVED** [review]

**Comment by kamleshbhalui** [comment]
@mbudiu-vmw Please merge it. is there any policy regarding who can have write access to p4lang/p4c repo ?

**Pull request merged**